### PR TITLE
♻️ Refactor/liquidations buy vault

### PR DESCRIPTION
**Pull request by Thomas-Smets**

### CODE REVIEW

File: src/Liquidator.sol [hunk 1]
```diff
@@ -69,176 +69,172 @@ contract Liquidator is Ownable {
...
-        (uint256 priceOfVault,, bool forSale) = getPriceOfVault(vaultAddress);
-
-        require(forSale, "LQ_BV: Not for sale");
+    function buyVault(address vault) external {
+        AuctionInformation memory auctionInformation_ = auctionInformation[vault];
+        require(auctionInformation_.inAuction, "LQ_BV: Not for sale");
 
-        address trustedCreditor = auctionInformation[vaultAddress].trustedCreditor;
-        address asset = auctionInformation[vaultAddress].baseCurrency;
+        uint256 priceOfVault = _calcPriceOfVault(auctionInformation_.startTime, auctionInformation_.openDebt);
```
Comment (j-vp) [inline src/Liquidator.sol:171]:
```suggestion
        uint256 priceOfVault = _calcPriceOfVault(block.timestamp - auctionInformation_.startTime, auctionInformation_.openDebt);
```

Comment (Thomas-Smets) [reply src/Liquidator.sol:171]: Don't think this is correct, see line 151

**Review by zekiblue — APPROVED** [review]

File: src/Factory.sol [hunk 2]
```diff
@@ -228,43 +250,22 @@ contract Factory is ERC721, FactoryGuardian {
...
-    /**
-     * @notice Helper transfer function that allows the contract to transfer ownership of the erc721.
-     * @dev This function is called by the contract when a vault is liquidated.
-     * This includes a transfer of ownership of the vault.
-     * We circumvent the ERC721 transfer function.
-     * @param vault Vault that needs to get transfered.
-     */
-    function _liquidateTransfer(address vault, address liquidator) internal {
-        address from = ownerOf[vaultIndex[vault]];
+        address from = ownerOf[vaultIndex[msg.sender]];
```
Comment (j-vp) [inline src/Factory.sol:259]: cache vault id:

```solidity
    function liquidate(address liquidator) external whenLiquidateNotPaused {
        require(isVault(msg.sender), "FTRY: Not a vault");
        uint256 id = vaultIndex[msg.sender];

        address from = ownerOf[id];
        unchecked {
            balanceOf[from]--;
            balanceOf[liquidator]++;
        }

        ownerOf[id] = liquidator;

        delete getApproved[id];
        emit Transfer(from, liquidator, id);
    }
```

File: src/test/7_Factory.t.sol [hunk 3]
```diff
@@ -663,82 +734,51 @@ contract FactoryTest is DeployArcadiaVaults {
...
         // Then: liquidate reverts
         vm.expectRevert("Guardian: liquidate paused");
-        vm.prank(liquidationInitiator);
-        factory.liquidate(address(proxy));
+        vm.prank(address(proxy));
+        factory.liquidate(liquidator_);
     }
 
-    function testSuccess_liquidate(address liquidationInitiator, uint128 openPosition) public {
+    function testSuccess_liquidate(address liquidator_, uint128 openPosition) public {
```
Comment (j-vp) [inline src/test/7_Factory.t.sol:767]: openposition isn't used

File: src/Liquidator.sol [hunk 4]
```diff
@@ -69,176 +69,172 @@ contract Liquidator is Ownable {
...
+    function calcLiquidationSettlementValues(uint256 openDebt, uint256 priceOfVault)
+        public
+        view
+        returns (uint256 badDebt, uint256 liquidationInitiatorReward, uint256 liquidationPenalty, uint256 remainder)
+    {
+        ClaimRatios memory claimRatios_ = claimRatios;
+
+        //openDebt is a uint128 -> all calculations can be unchecked
+        unchecked {
+            //Liquidation Initiator Reward is always payed out, independant of the final auction price
```
Comment (j-vp) [inline src/Liquidator.sol:217]:
```suggestion
            //Liquidation Initiator Reward is always paid out, independent of the final auction price
```

File: src/Liquidator.sol [hunk 5]
```diff
@@ -69,176 +69,172 @@ contract Liquidator is Ownable {
...
+            liquidationInitiatorReward = openDebt * claimRatios_.initiatorReward / 100;
+
+            //Final Auction price should at least cover the original debt and  Liquidation Initiator Reward.
```
Comment (j-vp) [inline src/Liquidator.sol:220]:
```suggestion
            //Final Auction price should at least cover the original debt and Liquidation Initiator Reward.
```

File: src/Vault.sol [hunk 6]
```diff
@@ -234,10 +235,10 @@ contract Vault {
     /////////////////////////////////////////////////////////////// */
 
     /**
-     * @notice Can be called by authorised applications to increase a margin position.
-     * @param baseCurrency_ The Base-currency in which the margin position is denominated
+     * @notice Called by trusted applications, checks if the Vault has sufficient free margin.
+     * @param baseCurrency_ The Base-currency in which the vault is denominated.
```
Comment (j-vp) [inline src/Vault.sol:239]:
```suggestion
     * @param baseCurrency_ The Base-currency in which the Vault is denominated.
```

File: src/Vault.sol [hunk 7]
```diff
@@ -331,32 +332,31 @@ contract Vault {
...
-    function liquidateVault(address liquidationInitiator) public onlyFactory returns (address liquidator_) {
-        uint256 usedMargin = getUsedMargin();
+    function liquidateVault(uint256 openDebt)
+        external
+        returns (address originalOwner, address baseCurrency_, address trustedCreditor_)
+    {
+        require(msg.sender == liquidator, "V_LV: You are not the liquidator");
 
-        require(getLiquidationValue() < usedMargin, "V_LV: This vault is healthy");
+        //In current Vault version, the vault can only have debt owed to a single creditor, the trustedCreditor
```
Comment (j-vp) [inline src/Vault.sol:349]:
```suggestion
        //In current Vault version, the Vault can only have debt owed to a single creditor, the trustedCreditor
```

File: src/Vault.sol [hunk 8]
```diff
@@ -331,32 +332,31 @@ contract Vault {
...
-        ILiquidator(liquidator).startAuction(
-            liquidationInitiator, owner, uint128(usedMargin), baseCurrency, trustedCreditor
-        );
+        //Transfer ownership of the ERC721 in Factory of the Vault to the Liquidator.
+        IFactory(IMainRegistry(registry).factoryAddress()).liquidate(msg.sender);
 
-        //Hook implemented on the trusted creditor contract to notify that the vault
-        //is being liquidated and trigger any necessary logic on the trustedCreditor.
-        ITrustedCreditor(trustedCreditor).liquidateVault(usedMargin);
+        //Transfer ownership of the Vault itself
```
Comment (j-vp) [inline src/Vault.sol:355]:
```suggestion
        //Transfer ownership of the Vault itself to the liquidator
```

File: src/Liquidator.sol [hunk 9]
```diff
@@ -69,176 +69,172 @@ contract Liquidator is Ownable {
...
+    function _calcPriceOfVault(uint256 startTime, uint256 openDebt) internal view returns (uint256 price) {
+        uint256 auctionTime = block.timestamp - startTime; //Can be unchecked
+
+        if (auctionTime > maxAuctionTime) {
             //ヽ༼ຈʖ̯ຈ༽ﾉ
-            totalPrice = 0;
+            price = 0;
         } else {
-            totalPrice = startPrice - priceDecrease;
+            price = uint256(openDebt) * startPriceMultiplier * (maxAuctionTime - auctionTime) / maxAuctionTime / 100;
```
[the remark this answers is not in the document]
Comment (Thomas-Smets) [reply src/Liquidator.sol:157]: So I keep it here?

**Pull request merged**